### PR TITLE
support semi-join for exists.

### PR DESCRIPTION
**Pull request by hanfei19910905**
@coocood @shenli @zimulala PTAL

### CODE REVIEW

File: plan/logical_plan_builder.go [hunk 1]
```diff
@@ -914,3 +925,43 @@ func (b *planBuilder) buildMaxOneRow(p LogicalPlan) LogicalPlan {
 	maxOneRow.correlated = p.IsCorrelated()
 	return maxOneRow
 }
+
+func deCorrelated(expr expression.Expression, outerPlan Plan) bool {
+	correlated := false
+	_, outer := extractColumn(expr, nil, nil)
```
Comment (coocood) [inline plan/logical_plan_builder.go]: rename to 'outerCols'.

File: plan/logical_plan_builder.go [hunk 2]
```diff
@@ -914,3 +925,43 @@ func (b *planBuilder) buildMaxOneRow(p LogicalPlan) LogicalPlan {
...
+		if outerPlan.GetSchema().GetIndex(c) != -1 {
+			c.Correlated = false
+		} else {
+			correlated = true
+		}
+	}
+	return correlated
+}
+
+func (b *planBuilder) buildSemiJoin(outerPlan, innerPlan LogicalPlan, onCondition []expression.Expression, inExpr bool) LogicalPlan {
```
Comment (coocood) [inline plan/logical_plan_builder.go]: When 'inExpr' would be set to false?

Comment (hanfei19910905) [inline plan/logical_plan_builder.go]: It won't be false currently, but next pr will use it.

File: executor/new_builder.go [hunk 3]
```diff
@@ -76,6 +76,30 @@ func (b *executorBuilder) buildJoin(v *plan.Join) Executor {
 	return e
 }
 
+func (b *executorBuilder) buildSemiJoin(v *plan.PhysicalHashSemiJoin) Executor {
```
Comment (shenli) [inline executor/new_builder.go]: Why not use buildHashSemiJoin? Is there any other type of semijoin?

Comment (hanfei19910905) [inline executor/new_builder.go]: yes,there exists sort merge semi join

**Comment by coocood** [comment]
LGTM

File: plan/logical_plan_builder.go [hunk 4]
```diff
@@ -914,3 +925,43 @@ func (b *planBuilder) buildMaxOneRow(p LogicalPlan) LogicalPlan {
 	maxOneRow.correlated = p.IsCorrelated()
 	return maxOneRow
 }
+
+func deCorrelated(expr expression.Expression, outerPlan Plan) bool {
```
Comment (shenli) [inline plan/logical_plan_builder.go]: Add comment for this function. I suggest changing its name to tryDecorrelated.

File: plan/logical_plan_builder.go [hunk 5]
```diff
@@ -894,6 +895,16 @@ func (b *planBuilder) buildApply(p, inner LogicalPlan, schema expression.Schema,
 }
 
 func (b *planBuilder) buildExists(p LogicalPlan) LogicalPlan {
+out:
+	for {
+		switch p.(type) {
+		case *Trim, *Projection, *NewSort, *Aggregation:
+			p = p.GetChildByIndex(0).(LogicalPlan)
+			p.SetParents()
```
Comment (shenli) [inline plan/logical_plan_builder.go]: This will set p's parents to empty slice?

File: plan/logical_plan_builder.go [hunk 6]
```diff
@@ -914,3 +925,44 @@ func (b *planBuilder) buildMaxOneRow(p LogicalPlan) LogicalPlan {
 	maxOneRow.correlated = p.IsCorrelated()
 	return maxOneRow
 }
+
+// tryDecorrelated try to remove the correlated column that can be find in the outerPlan's schema.
```
Comment (shenli) [inline plan/logical_plan_builder.go]: tries

**Comment by shenli** [comment]
LGTM

**Pull request merged**